### DISCUSSION TITLE
Single Package for All Platforms

### DISCUSSION THREAD

**Discussion by cliftonlabrum** [tekartik/sqflite#973] [Ideas]
First of all, this is a **_fantastic_** package. Thank you very much for creating and maintaining it!

I'm sorry if this is already explained somewhere and I missed it, but why not have a single `sqflite` package that relies on `sqlite3_flutter_libs` and works on all the platforms: **macOS**, **iOS**, **Android**, and **Windows**?

I know you have to use `sqflite_common_ffi` for Windows and set up a database factory, but that's always felt a little confusing to me, and involves extra steps (especially when deploying to Windows in production).

Why not have an all-in-one solution for all platforms? I'm genuinely curious. 😊

**Comment by alextekartik** [comment]
> why not have a single sqflite package that relies on sqlite3_flutter_libs and works on all the platforms: macOS, iOS, Android, and Windows?

Always including sqlite3_flutter_libs gives a larger binary size (about 2.5Mb on Android) even if you are not using it. It might not matter for you but some people don't need it and are fine with the platform bundled binary using `sqflite`

> Why not have an all-in-one solution for all platforms? I'm genuinely curious

Let me copy what I just wrote in: https://github.com/tekartik/sqflite/issues/948

Cross platform support is a bit weird, sorry about that. The `sqflite` package,
implemented as a "classic" flutter plugin supports only Android, iOS and MacOS.

Desktop support (Linux, Windows) is provided by the `sqflite_common_ffi` package.
It is a dart package so it also works on the dart VM (command line) so it is not
a flutter plugin. Its implementation uses ffi. This implementation can also
support Android, iOS and MacOS but it is not a flutter plugin.

Experimental web support is provided by the `sqflite_common_ffi_web` package.

> Is there a plan to publish a single federated package?

Not at this time. Some might prefer `sqflite_common_ffi` over `sqflite`. The first
giving you access to the latest sqlite3 version and the second giving a smaller binary size.
Some wants encryption support. It is hard to make the proper choice for everyone.

As a side note, I find it weird when creating a linux only flutter application, that
my app has to download the `win32` package. Currently there is no way to limit dependencies
to only the platform you are supporting. More dependencies means more chance to have breaking bugs 
that you cannot fix.

I would even be more inclined to publish a separate packages for each platform that people have to include
one by one for the platforms they want.

Also I'm a freelance developer, maintaining it in my free-time so I don't always have time and
the energy to support it (since 2017!) in the best way. Documentation is poor, I know. I'm not a native english speaker.
Sometimes issues like this one bring me the opportunity to improve the documentation so I will
likely copy this response to the documentation.

**Reply by cliftonlabrum** [reply]
Thank you for the thorough reply. That makes sense. 

Thanks for all you have done for the Flutter community! Also, your English is excellent. 😊

**Comment by rusty-snake** [comment]
> Always including sqlite3_flutter_libs gives a larger binary size (about 2.5Mb on Android) even if you are not using it. It might not matter for you but some people don't need it and are fine with the platform bundled binary using sqflite

No with sqflite_common_ffi 2.4.0 the sqlite.so is always included.

Is there a way to have sqflite setup in a way that uses system library on Android, does not bundle sqlite.so in Android builds. And use sqflite_common_ffi to run (for development purpose, not for production use) the same app under Linux?

**Comment by alextekartik** [comment]
Indeed that's an issue. I'm not aware of any selective way per platform. Might be a question for the flutter team or for the sqlite3 package. In the mean time having one app package per platform that includes what it needs is a (painful but robust) solution.

In an ideal world we could have a single package and a configuration to select native assets or plugins per platforms to allow android with sqflite or ffi without sqlite3 libs and linux with bundled libraries. That looks like a generic problem that many plugins might encounter. Mabye I missed something...any clue is welcome!
